### PR TITLE
[JENKINS-31698] Add support for folder hierarchies in ATH

### PR DESCRIPTION
**Pull request by andresrc**
[JENKINS-31698](https://issues.jenkins-ci.org/browse/JENKINS-31698)

Goals:
* Providing a mechanism to create jobs inside folders (including other folders).
* Review TopLevelItem equality contract in the presence of hierarchy.
* Create a simple test scenario involving a folder hierarchy.

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/test/acceptance/po/TopLevelItem.java [hunk 1]
```diff
@@ -56,14 +55,15 @@ public boolean equals(Object other) {
...
         
         // Equals and hashCode must be consistent with each other.
-        // Keep identity based on class and url, as in hashCode before.
-        // Shouldn't be url alone?
+        // We use URL only as the concrete class may not be detected, e.g. when obtaining a job
+        // from a BuildHistory
+        // TODO: equality checking in the presence of views.
         TopLevelItem rhs = (TopLevelItem) other;
-        return getClass().equals(rhs.getClass()) && Objects.equal(url, rhs.url);
+        return url.toExternalForm().equals(rhs.url.toExternalForm());
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/test/acceptance/po/TopLevelItem.java]: `URL.toString()` is calling `toExternalForm()` anyway, but ok.

**Comment by amuniz** [comment]
:bee:

**Comment by amuniz** [comment]
I think this fixes most of the current [ATH failures](https://jenkins.ci.cloudbees.com/job/core/job/acceptance-test-harness-stable/521/#showFailuresLink).

**Comment by andresrc** [comment]
@reviewbybees done

@olivergondza This PR fixes failures in the following tests:
* core.BuildHistoryTest.slave_build_history
* core.BuildHistoryTest.global_build_history
* core.BuildHistoryTest.view_build_history
* plugins.NodeLabelParameterPluginTest.run_on_online_slave_and_master_with_node_restriction
* plugins.NodeLabelParameterPluginTest.run_on_several_online_and_offline_slaves
* plugins.NodeLabelParameterPluginTest.run_on_several_slaves
* plugins.NodeLabelParameterPluginTest.trigger_if_succeeds_with_failed_main_build
* plugins.NodeLabelParameterPluginTest.pending_build_with_no_valid_node
* plugins.NodeLabelParameterPluginTest.run_on_a_particular_offline_slave

**Comment by olivergondza** [comment]
Thanks!

**Pull request merged**